File: problems/13_roman_to_Integer/roman_to_integer_C++/roman_to_integer.cpp
/**
 * LeetCode Problem 13: Roman to Integer
 *
 * This class provides a method to convert a Roman numeral string
 * into its corresponding integer value.
 *
 * The implementation handles standard Roman numeral notation,
 * including subtractive combinations (e.g., IV for 4, IX for 9, etc.).
 *
 * Example usage:
 *   Solution sol;
 *   int result = sol.romanToInt("MCMXCIV"); // result = 1994
 */

#include <string>

class Solution {
public:
    /**
     * Converts a Roman numeral string to its integer representation.
     *
     * @param s A string containing the Roman numeral (e.g., "MCMXCIV").
     * @return The integer value corresponding to the Roman numeral.
     */
    int romanToInt(std::string s)
    {
        int sum {0};

        // Lambda function to return the integer value of a Roman numeral character.
        auto romanToValue = [](char roman) constexpr {
            switch (roman)
            {
                case 'I': return 1;
                case 'V': return 5;
                case 'X': return 10;
                case 'L': return 50;
                case 'C': return 100;
                case 'D': return 500;
                case 'M': return 1000;
                default: return 0;
            }
        };

        // Iterate through the string, adding or subtracting values based on Roman numeral rules.
        for (int i = 0; i < s.length(); ++i)
        {
            if (i + 1 < s.length() && romanToValue(s[i]) < romanToValue(s[i + 1]))
            {
                // Subtractive combination detected, subtract current value.
                sum -= romanToValue(s[i]);
            }
            else
            {
                // Regular addition of numeral value.
                sum += romanToValue(s[i]);
            }
        }

        return sum;
    }
};

File: problems/13_roman_to_Integer/roman_to_integer_PYTHON/roman_to_integer.py
class Solution(object):
    """
    LeetCode Problem 13: Roman to Integer

    This class provides a method to convert a Roman numeral string
    into its corresponding integer value.

    The implementation handles standard Roman numeral notation,
    including subtractive combinations (e.g., IV for 4, IX for 9, etc.).
    """

    def romanToInt(self, s):
        """
        Converts a Roman numeral string to its integer representation.

        :param s: str - A string containing the Roman numeral (e.g., "MCMXCIV").
        :return: int - The integer value corresponding to the Roman numeral.

        Example:
        >>> sol = Solution()
        >>> sol.romanToInt("MCMXCIV")
        1994
        """

        # Mapping Roman numerals to their integer values
        d = {'I': 1, 'V': 5, 'X': 10, 'L': 50,
             'C': 100, 'D': 500, 'M': 1000}
        res = 0

        # Iterate through the string, adding or subtracting based on value comparison
        for i in range(len(s)):
            if i + 1 < len(s) and (d[s[i]] < d[s[i + 1]]):
                # Subtractive combination detected (e.g. IV, IX)
                res -= d[s[i]]
            else:
                # Regular addition of numeral value
                res += d[s[i]]

        return res


# Example usage for CI test
if __name__ == "__main__":
    sol = Solution()
    print(sol.romanToInt("MCMXCIV"))  # Output: 1994
